### PR TITLE
Multi DODAG support

### PR DESCRIPTION
**Pull request by laurentderu**
Currently Contiki only updates rank of parent belonging to the active DAG, parents belonging to other DAGs keep their initial rank even when receiving DIO from them.

This PR adds a mechanism to update the rank of these parents and also prevent the DODAG Root to join another DAG. With #912, a node can be part of MAX_DODAG_PER_INSTANCE DODAGs and drops dead ones.

### CODE REVIEW

File: core/net/rpl/rpl-dag.c [hunk 1]
```diff
@@ -741,6 +741,7 @@ rpl_select_parent(rpl_dag_t *dag)
 
   if(best != NULL) {
     rpl_set_preferred_parent(dag, best);
+    dag->rank = dag->instance->of->calculate_rank(dag->preferred_parent, 0);
```
Comment (simonduq) [inline core/net/rpl/rpl-dag.c]: Don't we also need to set the dag rank to infinite if there is no best parent?

File: core/net/rpl/rpl-dag.c [hunk 2]
```diff
@@ -1023,8 +1024,9 @@ rpl_add_dag(uip_ipaddr_t *from, rpl_dio_t *dio)
   PRINTF("\n");
 
   ANNOTATE("#A join=%u\n", dag->dag_id.u8[sizeof(dag->dag_id) - 1]);
-
-  rpl_process_parent_event(instance, p);
+  if(dag->rank != ROOT_RANK(instance)) {
+    rpl_process_parent_event(instance, p);
+  }
```
Comment (simonduq) [inline core/net/rpl/rpl-dag.c]: Why is this needed? A comment explaining this would be good!

File: core/net/rpl/rpl-dag.c [hunk 3]
```diff
@@ -1282,6 +1302,9 @@ rpl_process_dio(uip_ipaddr_t *from, rpl_dio_t *dio)
     }
   }
 
+  //Parent info has been updated, trigger rank recalculation
```
Comment (simonduq) [inline core/net/rpl/rpl-dag.c]: please use `/**/` comments instead

File: core/net/rpl/rpl-dag.c [hunk 4]
```diff
@@ -1100,6 +1107,19 @@ rpl_recalculate_ranks(void)
     }
     p = nbr_table_next(rpl_parents, p);
   }
+  if (!updated) {
```
Comment (simonduq) [inline core/net/rpl/rpl-dag.c]: the extra space after `if` should be removed

Comment (simonduq) [inline core/net/rpl/rpl-dag.c]: I don't understand what this does. When no parent needed update, we update the preferred parent of every instance and dag? Clarification needed here

File: core/net/rpl/rpl-dag.c (hunk 1 above)
Comment (laurentderu) [inline core/net/rpl/rpl-dag.c]: Indeed, that is missing, I will add it

File: core/net/rpl/rpl-dag.c (hunk 2 above)
Comment (laurentderu) [inline core/net/rpl/rpl-dag.c]: rpl_process_parent_event() update the rank of the node, assuming it's a RPL router and not a Root. All the other calls to rpl_process_parent_event() are either protected the same way (rpl_process_dio() or global_repair()) or is never called (rpl_recalculate_ranks())

I will add a comment

File: core/net/rpl/rpl-dag.c (hunk 4 above)
Comment (laurentderu) [inline core/net/rpl/rpl-dag.c]: The PR and its commits are quite old, I believe this code was necessary when rpl_parents table did not exist and only the current dag parents were updated.

File: core/net/rpl/rpl-dag.c (hunk 2 above)
Comment (laurentderu) [inline core/net/rpl/rpl-dag.c]: Actually no longer needed since fd97136fee495ac0a261dbff642423a45db0ba0c

File: core/net/rpl/rpl-dag.c (hunk 4 above)
Comment (laurentderu) [inline core/net/rpl/rpl-dag.c]: I confirm, this is no longer needed, all the parents are now checked.

**Comment by laurentderu** [comment]
PR updated and obsolete code removed. I let it run on during this week-end and network seems stable and multi-dodag still working.

**Comment by simonduq** [comment]
:+1:

**Comment by nvt** [comment]
:+1:
